src/dcp/dcp.c

```diff
@@ -346,6 +359,291 @@ static int daos_any_error(int rank, bool local_daos_error, int flag_daos_args)
     return 0;
 }
 
+static int daos_copy_recx_single(daos_key_t *dkey,
+                            daos_handle_t *src_oh,
+                            daos_handle_t *dst_oh,
+                            daos_iod_t *iod)
+{
+    /* if iod_type is single value just fetch iod size from source
+     * and update in destination object */
+    int buf_len = (int)(*iod).iod_size;
+    char buf[buf_len];
+    d_sg_list_t sgl;
+    d_iov_t iov;
+    int rc;
+
+    /* set sgl values */
+    sgl.sg_nr     = 1;
+    sgl.sg_nr_out = 0;
+    sgl.sg_iovs   = &iov;
+    d_iov_set(&iov, buf, buf_len);
+    rc = daos_obj_fetch(*src_oh, DAOS_TX_NONE, 0, dkey, 1, iod, &sgl, NULL, NULL);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "DAOS object fetch returned with errors: ", MFU_ERRF,
+	        MFU_ERRP(-MFU_ERR_DAOS));
+    }
+    rc = daos_obj_update(*dst_oh, DAOS_TX_NONE, 0, dkey, 1, iod, &sgl, NULL);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "DAOS object fetch returned with errors: ", MFU_ERRF,
+	        MFU_ERRP(-MFU_ERR_DAOS));
+    }
+    return rc;
+}
+
+static int daos_copy_recx_array(daos_key_t *dkey,
+                           daos_key_t *akey,
+                           daos_handle_t *src_oh,
+                           daos_handle_t *dst_oh,
+                           daos_iod_t *iod)
+{
+    daos_anchor_t recx_anchor = {0}; 
+    int rc;
+    int i;
+    while (!daos_anchor_is_eof(&recx_anchor)) {
+        daos_epoch_range_t	eprs[5];
+        daos_recx_t		recxs[5];
+        daos_size_t		size;
+
+        /* list all recx for this dkey/akey */
+        uint32_t number = 5;
+        rc = daos_obj_list_recx(*src_oh, DAOS_TX_NONE, dkey, akey,
+	                        &size, &number, recxs, eprs,
+				&recx_anchor, true, NULL);
+
+        /* if no recx is returned for this dkey/akey move on */
+        if (number == 0) 
+            continue;
+
+        for (i = 0; i < number; i++) {
+            uint64_t    buf_len = recxs[i].rx_nr;
+            char        buf[buf_len];
+            d_sg_list_t sgl;
+            d_iov_t     iov;
+
+            /* set iod values */
+            (*iod).iod_type  = DAOS_IOD_ARRAY;
+            (*iod).iod_size  = 1;
+            (*iod).iod_nr    = 1;
+            (*iod).iod_recxs = &recxs[i];
+
+            /* set sgl values */
+            sgl.sg_nr     = 1;
```

Comment:
@daltonbohning it doesn't quite work like that since sgl just tells you how many segments will be filled, but you can specify the number you want to fetch in the API call. I can look into doing it that way.

Reply:
@dsikich Anyway, it would just be a performance optimization if it's possible, so we can look into it for a future iteration if needed.